sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,357 @@
+    }
+  }
+
+  /**
+   * Register the outbound data logical endpoint, returns the FnDataReceiver for processing the
+   * endpoint's outbound data.
+   */
+  public <T> FnDataReceiver<T> registerOutputDataLocation(String pTransformId, Coder<T> coder) {
+    if (outputDataReceivers.containsKey(pTransformId)) {
+      throw new IllegalStateException(
+          "Outbound data endpoint already registered for " + pTransformId);
+    }
+    Receiver<T> receiver = new Receiver<>(coder, totalByteCounter);
+    if (timeLimit > 0) {
+      synchronized (flushLock) {
+        outputDataReceivers.put(pTransformId, receiver);
+        return data -> {
+          checkFlushThreadException();
+          synchronized (flushLock) {
+            receiver.accept(data);
+            if (totalByteCounter.longValue() > sizeLimit) {
+              flush();
+            }
+          }
+        };
+      }
+    }
+    outputDataReceivers.put(pTransformId, receiver);
+    return data -> {
+      receiver.accept(data);
+      if (totalByteCounter.longValue() > sizeLimit) {
+        flush();
+      }
+    };
+  }
+
+  /**
+   * Register the outbound timers logical endpoint, returns the FnDataReceiver for processing the
+   * endpoint's outbound timers data.
+   */
+  public <T> FnDataReceiver<T> registerOutputTimersLocation(
+      String pTransformId, String timerFamilyId, Coder<T> coder) {
+    KV<String, String> timerKey = KV.of(pTransformId, timerFamilyId);
+    if (outputTimersReceivers.containsKey(timerKey)) {
+      throw new IllegalStateException(
+          "Outbound timers endpoint already registered for " + timerKey);
+    }
+    Receiver<T> receiver = new Receiver<>(coder, totalByteCounter);
+    if (timeLimit > 0) {
+      synchronized (flushLock) {
+        outputTimersReceivers.put(timerKey, receiver);
+        return timers -> {
+          checkFlushThreadException();
+          synchronized (flushLock) {
+            receiver.accept(timers);
+            if (totalByteCounter.longValue() > sizeLimit) {
+              flush();
+            }
+          }
+        };
+      }
+    }
+    outputTimersReceivers.put(timerKey, receiver);
+    return timers -> {
+      receiver.accept(timers);
+      if (totalByteCounter.longValue() > sizeLimit) {
+        flush();
+      }
+    };
+  }
+
+  public void flush() throws IOException {
+    Elements.Builder elements = convertBufferForTransmission();
+    if (elements.getDataCount() > 0 || elements.getTimersCount() > 0) {
+      outboundObserver.onNext(elements.build());
+    }
+  }
+
+  /**
+   * Closes the streams for all registered outbound endpoints. Should be called at the end of each
+   * bundle.
+   */
+  public void sendBufferedDataAndFinishOutboundStreams() {
+    if (totalByteCounter.longValue() == 0
+        && outputTimersReceivers.isEmpty()
+        && outputDataReceivers.isEmpty()) {
+      return;
+    }
+    Elements.Builder bufferedElements = convertBufferForTransmission();
+    LOG.debug(
```

Comment:
If you implement Receiver#toString() you can do something like
```
LOG.debug("Closing streams for instruction {} for outbound data {} and timers {}.", processBundleInstructionIdSupplier.get(), outputDataReceivers, outputTimersReceivers);
```

You can get pretty printing for the timers by creating something like TimerEndpoint instead of a KV and implement toString on it.